Source/Particles/Collision/BackgroundMCCCollision.cpp

```diff
@@ -0,0 +1,376 @@
+              );
+        if (nu > nu_max) {
+            nu_max = nu;
+        }
+    }
+    return nu_max;
+}
+
+void
+BackgroundMCCCollision::doCollisions (amrex::Real cur_time, MultiParticleContainer* mypc)
+{
+    WARPX_PROFILE("BackgroundMCCCollision::doCollisions()");
+
+    const amrex::Real dt = WarpX::GetInstance().getdt(0);
+    if ( int(std::floor(cur_time/dt)) % m_ndt != 0 ) return;
+
+    auto& species1 = mypc->GetParticleContainerFromName(m_species_names[0]);
+    // this is a very ugly hack to have species2 be a reference and be
+    // defined in the scope of doCollisions
+    auto& species2 = (
+                      (m_species_names.size() == 2) ?
+                      mypc->GetParticleContainerFromName(m_species_names[1]) :
+                      mypc->GetParticleContainerFromName(m_species_names[0])
+                      );
+
+    if (!init_flag) {
+        m_mass1 = species1.getMass();
+
+        // calculate maximum collision frequency without ionization
+        m_nu_max = get_nu_max(m_scattering_processes);
+
+        // calculate total collision probability
+        auto coll_n = m_nu_max * dt;
+        m_total_collision_prob = 1.0 - std::exp(-coll_n);
+
+        // dt has to be small enough that a linear expansion of the collision
+        // probability is sufficiently accurately, otherwise the MCC results
+        // will be very heavily affected by small changes in the timestep
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(coll_n < 0.1,
+            "dt is too large to ensure accurate MCC results"
+        );
+
+        if (ionization_flag) {
+            // calculate maximum collision frequency for ionization
+            m_nu_max_ioniz = get_nu_max(m_ionization_processes);
+
+            // calculate total ionization probability
+            auto coll_n_ioniz = m_nu_max_ioniz * dt;
+            m_total_collision_prob_ioniz = 1.0 - std::exp(-coll_n_ioniz);
+
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(coll_n_ioniz < 0.1,
+                "dt is too large to ensure accurate MCC results"
+            );
+
+            // if an ionization process is included the secondary species mass
+            // is taken as the background mass
+            m_background_mass = species2.getMass();
+        }
+        // if no neutral species mass was specified and ionization is not
+        // included assume that the collisions will be with neutrals of the
+        // same mass as the colliding species (like with ion-neutral collisions)
+        else if (m_background_mass == -1) {
+            m_background_mass = species1.getMass();
+        }
+
+        amrex::Print() <<
+            "Setting up collisions for " << m_species_names[0] << " with total "
+            "collision probability: " <<
+            m_total_collision_prob + m_total_collision_prob_ioniz << "\n";
+
+        init_flag = true;
+    }
+
+    // Loop over refinement levels
+    for (int lev = 0; lev <= species1.finestLevel(); ++lev) {
```

Comment:
Potentially, not performance critical:
```suggestion
    for (int lev = 0, auto const flvl = species1.finestLevel(); lev <= flvl; ++lev) {
```